### DISCUSSION TITLE
A new package

### DISCUSSION THREAD

**Discussion by kenneth-gray** [kenneth-gray/data-mocks-server#65] [Ideas]
It's time to rethink `data-mocks-server`:

- Name change to `scenario-mock-server` (https://github.com/kenneth-gray/data-mocks-server/issues/54)
- Single scenario selection (https://github.com/kenneth-gray/data-mocks-server/issues/63)
- Allow a scenario to be the default (https://github.com/kenneth-gray/data-mocks-server/issues/63)
- No more default parameter (https://github.com/kenneth-gray/data-mocks-server/issues/63)
- `id`, `name`, `description` (https://github.com/kenneth-gray/data-mocks-server/issues/46)
- Extend scenario capability (https://github.com/kenneth-gray/data-mocks-server/issues/60)
- Change shape of response for REST mocks
- Overhaul GQL mocking
- Schema checking for REST
- Schema checking GQL

Let's quote each part so we can discuss in threads each point.

**Comment by kenneth-gray** [comment]
> Name change to `scenario-mock-server` (https://github.com/kenneth-gray/data-mocks-server/issues/54)

**Reply by kenneth-gray** [reply]
Done

**Comment by kenneth-gray** [comment]
> Single scenario selection (https://github.com/kenneth-gray/data-mocks-server/issues/63)

**Reply by freddybushboy** [reply]
This and the "Allow scenario to be the default" are really great changes IMO. At the moment we have a real lack of consistency in our scenarios because it's not clear how scenarios are inheriting from one another. 

Scenarios ideally represent a consistent backend state across multiple endpoints. With the fact that all scenarios inherit the "default" mocks can lead to inconsistencies here. To give a ludicrous example:

```
# default scenario
- Species endpoint => 'duck'
- Feet endpoint => 'webbed'

# pig scenario
- Species endpoint => 'pig'
```

It's possible what you wanted to test when you created this scenario is covered by this, and only the species is important to you, but these scenarios are available for people to use and whether you care about it or no, your pig now has webbed feet.

To then add on the fact that you can select multiple scenarios that may or may not define responses for each endpoint it becomes unclear what you will get, but it's unlikely to be consistent. 

Always selecting a single scenario removes this second issue but it does add the mean you have to define ALL responses you want to mock for each scenario - this can be made easier with the "Extend scenario capability" stuff, which can be discussed below.

**Reply by kenneth-gray** [reply]
Single scenario selection now exists in https://github.com/kenneth-gray/scenario-mock-server, but I decided not to allow the `initiallySelectedScenario` to be a feature - it's just the first declared one and that's it.

**Comment by kenneth-gray** [comment]
> Allow a scenario to be the default (https://github.com/kenneth-gray/data-mocks-server/issues/63)

**Reply by freddybushboy** [reply]
Main point on the current "default" scenario isn't descriptive so its not clear what kind of user is represented, and people add response here pretty willy nilly as new endpoints arrive without considering if they make sense with the other responses in this scenario. To use my silly example from the other thread:

People will add new endpoints to default, without considering that the stuff in here is pretty duck-like. keep in mind there would be many more endpoints to its not as easy to scan as this.
```
# default scenario
- Species endpoint => 'duck'
- Feet endpoint => 'webbed'

# pig scenario
- Species endpoint => 'pig'
```

It's clearer what's going on here:
```
# duck scenario
- Species endpoint => 'duck'
- Feet endpoint => 'webbed'

# pig scenario
- Species endpoint => 'pig'
- Feet endpoint => 'hooved'

defaultScenario = "duck"
```

**Reply by kenneth-gray** [reply]
As mentioned in https://github.com/kenneth-gray/data-mocks-server/discussions/65#discussioncomment-5285676, this hasn't made it into the new package.

**Comment by kenneth-gray** [comment]
> No more default parameter (https://github.com/kenneth-gray/data-mocks-server/issues/63)

**Reply by freddybushboy** [reply]
I've touched on this in other comments as its sort of all interconnected. [See here for example](https://github.com/kenneth-gray/data-mocks-server/discussions/65#discussioncomment-4861380)

**Reply by kenneth-gray** [reply]
Done

**Comment by kenneth-gray** [comment]
> `id`, `name`, `description` (https://github.com/kenneth-gray/data-mocks-server/issues/46)

**Reply by kenneth-gray** [reply]
- `id`s would be used for the API endpoints
  - required
- `name`s would be used in the UI as a concise description
  - required
- `descriptions`s would be used as longer descriptions in case having some context is helpful
  - optional

**Reply by kenneth-gray** [reply]
Done

**Comment by kenneth-gray** [comment]
> Extend scenario capability (https://github.com/kenneth-gray/data-mocks-server/issues/60)

**Reply by freddybushboy** [reply]
Following on from [my comments above](https://github.com/kenneth-gray/data-mocks-server/discussions/65#discussioncomment-4861380) I think this feature is very helpful if these other changes are implemented. 

The way scenarios extend from one another and the default mocks is unpredictable - deliberately extending scenarios at least makes it clear what the person who created the scenario was intending. It's not fool proof though, you could still make inconsistent scenarios but I think this extending system would encourage you to make some solid base scenarios for the major types of scenario you support, an extend those resulting in better scenarios.

**Reply by kenneth-gray** [reply]
I'd be interested in thoughts on extending multiple scenarios as well. At least this time if you did that the ordering of the scenarios being extended would be consistent (instead of the current random it is now).

**Reply by freddybushboy** [reply]
Personally I'd start by not having the ability to extend multiple just purely for simplicity and an MVP approach. I wonder how many times you'd do this and if it encourages you to make incomplete scenarios purely to extend from. 

To use a silly example again - the `red` scenario probably isn't really a proper scenario here and maybe just defines a single endpoint response?

```
// redPig scenario
Extends 'pig' and 'red'
```

It would still appear in the list of scenarios even though a dev maybe just made it to extend from. Not **necessarily a bad thing**, but I do think its a layer of complexity that might be better to wait and see if its necessary?
 
 Of course, only being able to extend from a single scenario doesn't actually prevent this, but maybe makes it less likely?

**Reply by kenneth-gray** [reply]
@Samuel-West Keen to get your thoughts on this one as well as you've used the library historically too (check out the top level discussion for other potential changes)

**Reply by kenneth-gray** [reply]
Done.

**Comment by kenneth-gray** [comment]
> Change shape of response for REST mocks

**Reply by kenneth-gray** [reply]
Currently the response looks roughly like like the following:

```
response?: unknown | Function;
responseCode?: number;
responseHeaders?: Record<string, string>;
responseDelay?: number;
```

`response` can take a function which either returns a value or a special object using `__override`, like so:

```js
const mock = {
  url: '/some-url',
  method: 'GET',
  response: ({ body }) => {
    if (body.name === 'error1') {
      return {
        __override: {
          response: { message: 'something went wrong' },
          responseCode: 400,
          responseDelay: 1000,
        },
      };
    }

    if (body.name === 'error2') {
      return {
        __override: {
          response: { message: 'something else went wrong' },
          responseCode: 500,
          responseDelay: 2000,
        },
      };
    }

    if (body.name === 'notFound') {
      return {
        __override: {
          response: { message: 'no data here' },
          responseCode: 404,
        },
      };
    }

    // No __override necessary, this is the response on success
    return { message: 'success' };
  },
};
```

I'd like to change this so it looks like the following instead:

```js
const mock = {
  url: '/some-url',
  method: 'GET',
  response: ({ body }) => {
    if (body.name === 'error1') {
      return {
        status: 400,
        delay: 1000,
        data: { message: 'something went wrong' },
      };
    }

    if (body.name === 'error2') {
      return {
        status: 500,
        delay: 2000,
        data: { message: 'something else went wrong' },
      };
    }

    if (body.name === 'notFound') {
      return {
        status: 404,
        data: { message: 'no data here' },
      };
    }

    // status would default to 200, headers to {}, delay to 0
    return { data: { message: 'success' } };
  },
};
```

The new mock would look roughly like the following:

```
response: {
  status?: number;
  headers?: Record<string, string>;
  data?: unknown;
  delay?: number;
} | Function;
```

**Reply by kenneth-gray** [reply]
This is now in version 10 of data-mocks-server

**Reply by kenneth-gray** [reply]
And ported to https://github.com/kenneth-gray/scenario-mock-server

**Comment by kenneth-gray** [comment]
> Overhaul GQL mocking

**Reply by kenneth-gray** [reply]
Apollo has a much better way of default mocking GQL types, the new library should try and match this as an alternative to checking for the `operationName`.

**Reply by kenneth-gray** [reply]
It should also be easier to respond under `data` or `error`. It feels a bit error prone at the moment.

**Reply by kenneth-gray** [reply]
Turns out mock service worker uses the `operationName` approach - keen to get opinions.

**Reply by kenneth-gray** [reply]
> It should also be easier to respond under `data` or `error`. It feels a bit error prone at the moment.

Types have been changed to make this better in the TS world.

**Reply by kenneth-gray** [reply]
This is on hold for now.

**Comment by kenneth-gray** [comment]
> Schema checking for REST

**Reply by kenneth-gray** [reply]
Allow for Open API Specifications - as a separate parameter?

**Reply by kenneth-gray** [reply]
Should mismatches actually error the API or print to stderr?

**Reply by kenneth-gray** [reply]
Question about the amount of static analysis that can be done due to responses allowing functions (so context can be used, etc.)

**Reply by kenneth-gray** [reply]
New feature, not for MVP

**Comment by kenneth-gray** [comment]
> Schema checking GQL

**Reply by kenneth-gray** [reply]
Allow for GQL schemas, then some similar questions as https://github.com/kenneth-gray/data-mocks-server/discussions/65#discussioncomment-4861182 for errors and catching them in the first place.

**Reply by kenneth-gray** [reply]
New feature, not for MVP
